From: Hassan Shahbazi <h-shahbazi@hotmail.com>
Date: Mon, 15 Feb 2021 20:42:10 +0200
Subject: [19aa2fa32708] [PATCH 01/02] task 08: debugs for 'jiffies'

Signed-off-by: Hassan Shahbazi <h-shahbazi@hotmail.com>
---
 task_08/ktask08.c |  2 ++
 task_08/script.sh | 10 +++++++---
 2 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/task_08/ktask08.c b/task_08/ktask08.c
index 3a5b448..545a44b 100644
--- a/task_08/ktask08.c
+++ b/task_08/ktask08.c
@@ -5,6 +5,7 @@
 #include <linux/debugfs.h>
 #include <linux/fs.h>
 #include <linux/slab.h>
+#include <linux/jiffies.h>
 
 MODULE_AUTHOR("Hassan Shahbazi");
 MODULE_DESCRIPTION("Linux Kernel Mentorship task 08");
@@ -57,6 +58,7 @@ static int __init task_init(void)
 {
 	eudyptula = debugfs_create_dir(DIRECTORY, NULL);
 	debugfs_create_file("id", 0755, eudyptula, NULL, &fops);
+	debugfs_create_u64("jiffies", 0444, eudyptula, (u64*)&jiffies);
 	return 0;
 }
 module_init(task_init);
diff --git a/task_08/script.sh b/task_08/script.sh
index c8133e1..981a404 100755
--- a/task_08/script.sh
+++ b/task_08/script.sh
@@ -1,9 +1,13 @@
 #! /bin/sh
 
-echo "read from the device"
+echo "read from id"
 cat /sys/kernel/debug/eudyptula/id
-
 echo ""
-echo "write to the device"
+
+echo "write to the id"
 echo 19aa2fa32708 > /sys/kernel/debug/eudyptula/id
 echo "Succeed"
+
+echo "read from jiffies"
+cat /sys/kernel/debug/eudyptula/jiffies
+echo ""
-- 
2.26.2

